File: Set 7/World Ladder 1/Ladder.cpp
#include <iostream>
#include <unordered_set>
#include <queue>
#include <string>
#include <iterator>
using namespace std;

int ladder(string start, string target, unordered_set<string>& D) {
    if (start == target) {
        return 0;
    }
    if (D.find(target) == D.end()) {
        return 0;
    }

    int level = 0;
    int wordLength = start.length();
    queue<string> Q;
    Q.push(start);

    while (!Q.empty()) {
        level++;
        int size = Q.size();

        for (int i = 0; i < size; i++) {
            string word = Q.front();
            Q.pop();

            for (int pos = 0; pos < wordLength; pos++) {
                char origChar = word[pos];

                for (char c = 'a'; c <= 'z'; c++) {
                    word[pos] = c;

                    if (word == target) {
                        return level + 1;
                    }

                    if (D.find(word) != D.end()) {
                        D.erase(word);
                        Q.push(word);
                    }
                }

                word[pos] = origChar;
            }
        }
    }

    return 0;
}

int main() {
    int t;
    cin >> t;

    while (t--) {
        unordered_set<string> D;
        string word;
        string start, end;
        
        while (cin >> word) {
            if (word == "-1") break;
            D.insert(word);
        }
        
        cin >> start >> end;
        
        cout << ladder(start, end, D) << endl;
    }

    return 0;
}

File: Set 7/World Ladder 1/Ladder.py
from collections import deque

def ladder(start, target, D):
    if start == target:
        return 0
    if target not in D:
        return 0

    level, wordlength = 0, len(start)
    Q = deque()
    Q.append(start)

    while len(Q) > 0:
        level += 1
        sizeofQ = len(Q)

        for i in range(sizeofQ):
            word = [j for j in Q.popleft()]

            for pos in range(wordlength):
                orig_char = word[pos]

                for c in range(ord('a'), ord('z') + 1):
                    word[pos] = chr(c)

                    if "".join(word) == target:
                        return level + 1

                    if "".join(word) not in D:
                        continue
                        
                    D.remove("".join(word))  
                    Q.append("".join(word))

                word[pos] = orig_char

    return 0

t = int(input())
for _ in range(t):
    strings = set(input().split())
    start = input()
    end = input()
    print(ladder(start, end, strings))
